File: android/src/main/java/com/cap/plugins/statusbar/CapacitorStatusBarPlugin.java
package com.cap.plugins.statusbar;

import android.view.View;

import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "CapacitorStatusBar")
public class CapacitorStatusBarPlugin extends Plugin {
    private final CapacitorStatusBar implementation = new CapacitorStatusBar();

    @Override
    public void load() {
        super.load();
        getActivity().runOnUiThread(() -> {
            View webView = getBridge().getWebView();
            implementation.ensureEdgeToEdgeConfigured(getActivity(), webView);
            implementation.applyDefaultStyle(getActivity());
        });
    }

    @PluginMethod
    public void setStyle(PluginCall call) {
        try {
            String style = call.getString("style");
            String color = call.getString("color");
            if (style == null) {
                call.reject("style is required");
                return;
            }
            getActivity().runOnUiThread(() -> {
                implementation.setStyle(getActivity(), style, color);
                call.resolve();
            });
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void show(PluginCall call) {
        try {
            getActivity().runOnUiThread(() -> {
                implementation.showStatusBar(getActivity(), true);
                call.resolve();
            });
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void hide(PluginCall call) {
        try {
            getActivity().runOnUiThread(() -> {
                implementation.hideStatusBar(getActivity());
                call.resolve();
            });
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void setOverlaysWebView(PluginCall call) {
        try {
            Boolean value = call.getBoolean("value");
            if (value == null) {
                call.reject("value is required");
                return;
            }
            getActivity().runOnUiThread(() -> {
                implementation.setOverlaysWebView(getActivity(), value);
                call.resolve();
            });
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void showNavigationBar(PluginCall call) {
        try {
            getActivity().runOnUiThread(() -> {
                implementation.showNavigationBar(getActivity(), true);
                call.resolve();
            });
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void hideNavigationBar(PluginCall call) {
        try {
            getActivity().runOnUiThread(() -> {
                implementation.hideNavigationBar(getActivity(), "slide");
                call.resolve();
            });
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void getSafeAreaInsets(PluginCall call) {
        try {
            getActivity().runOnUiThread(() -> {
                java.util.Map<String, Integer> insets = implementation.getSafeAreaInsets(getActivity());
                com.getcapacitor.JSObject result = new com.getcapacitor.JSObject();
                result.put("top", insets.get("top"));
                result.put("bottom", insets.get("bottom"));
                result.put("left", insets.get("left"));
                result.put("right", insets.get("right"));
                call.resolve(result);
            });
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }
}

File: ios/Sources/CapacitorStatusBarPlugin/CapacitorStatusBarPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(CapacitorStatusBarPlugin)
public class CapacitorStatusBarPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CapacitorStatusBarPlugin"
    public let jsName = "CapacitorStatusBar"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "setStyle", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "show", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "hide", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setOverlaysWebView", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "showNavigationBar", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "hideNavigationBar", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getSafeAreaInsets", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = CapacitorStatusBar()

    override public func load() {
        super.load()
        implementation.setWebView(bridge?.webView)
        // Apply default style based on system theme on plugin load
        implementation.applyDefaultStyle()
    }

    @objc func setStyle(_ call: CAPPluginCall) {
        guard let style = call.getString("style") else {
            call.reject("style is required")
            return
        }
        let color = call.getString("color")
        implementation.setStyle(style: style, colorHex: color) {
            call.resolve()
        }
    }

    @objc func show(_ call: CAPPluginCall) {
        implementation.show(animated: true) {
            call.resolve()
        }
    }

    @objc func hide(_ call: CAPPluginCall) {
        implementation.hide(animation: "slide") {
            call.resolve()
        }
    }

    @objc func setOverlaysWebView(_ call: CAPPluginCall) {
        guard let value = call.getBool("value") else {
            call.reject("value is required")
            return
        }
        implementation.setOverlaysWebView(value: value) {
            call.resolve()
        }
    }

    @objc func showNavigationBar(_ call: CAPPluginCall) {
        implementation.showNavigationBar(animated: true) {
            call.resolve()
        }
    }

    @objc func hideNavigationBar(_ call: CAPPluginCall) {
        implementation.hideNavigationBar(animation: "slide") {
            call.resolve()
        }
    }

    @objc func getSafeAreaInsets(_ call: CAPPluginCall) {
        implementation.getSafeAreaInsets { insets in
            call.resolve([
                "top": insets["top"] ?? 0,
                "bottom": insets["bottom"] ?? 0,
                "left": insets["left"] ?? 0,
                "right": insets["right"] ?? 0
            ])
        }
    }
}
